Google.Api.Gax.Grpc/ForwardingCallInvoker.cs

```diff
@@ -0,0 +1,155 @@
+        /// <typeparam name="TTargetRequest">The type of the target request, i.e. the request we'll forward on
+        /// to <paramref name="originalInvoker"/>.</typeparam>
+        /// <typeparam name="TTargetResponse">The type of the target response, i.e. the response we expect to be
+        /// returned by <paramref name="originalInvoker"/>.</typeparam>
+        /// <param name="originalInvoker">The original invoker that will handle the request.</param>
+        /// <param name="sourceMethodFullName">The full name (as reported by <see cref="Method{TRequest, TResponse}.FullName"/>)
+        /// of the method to forward.</param>
+        /// <param name="targetMethod">The target method to call on <paramref name="originalInvoker"/>.</param>
+        /// <param name="requestConverter">A delegate to convert source requests to target requests.</param>
+        /// <param name="responseConverter">A delegate to convert target responses to source responses, with
+        /// additional context being provided from the original source request.</param>
+        /// <returns>A call invoker forwarding the specified call.</returns>
+        public static CallInvoker Create<TSourceResponse, TTargetRequest, TTargetResponse>(
+            CallInvoker originalInvoker, string sourceMethodFullName, Method<TTargetRequest, TTargetResponse> targetMethod,
+            Func<TSourceRequest, TTargetRequest> requestConverter,
+            Func<TSourceRequest, TTargetResponse, TSourceResponse> responseConverter)            
+            where TSourceResponse : class
+            where TTargetRequest : class
+            where TTargetResponse : class =>
+            new ForwardingCallInvoker<TSourceRequest, TSourceResponse, TTargetRequest, TTargetResponse>(originalInvoker, sourceMethodFullName, targetMethod, requestConverter, responseConverter);
+    }
+#endif
+
+    /// <summary>
+    /// A <see cref="CallInvoker"/> which forwards specific calls to an existing invoker,
+    /// transforming the requests and responses as necessary.
+    /// </summary>
+    /// <remarks>
+    /// <para>
+    /// It would be cleaner to write an interceptor for this functionality, but that doesn't allow for the request/response types to be changed.
+    /// </para>
+    /// <para>
+    /// Currently, only unary methods are supported, and only a single method can be forwarded. Any
+    /// other method results in an <see cref="RpcException"/> with a status code of <see cref="StatusCode.Unimplemented"/>.
+    /// The type parameters of this class would make it hard to support multiple calls, which is why the factory
+    /// class method has a return type of <see cref="CallInvoker"/> rather than the concrete class: we may implement
+    /// multi-method support via composition of multiple call invokers.
+    /// </para>
+    /// </remarks>
+    internal sealed class ForwardingCallInvoker<TSourceRequest, TSourceResponse, TTargetRequest, TTargetResponse> : CallInvoker
```

Comment:
nit: Any reason why this one is not inside the `#if REGAPIC`?

Reply:
Yes - it makes testing a lot simpler. Basically, the only things we've got in `#if REGAPIC` are the public entry points for the REGAPIC work. Everything else is internal - so we're shipping extra unused code in our current GA releases, but that's not really a big problem - the important point is that we can still change the API.

Reply:
Ok